common/persistence/wrappers/metered/base.go

```diff
@@ -0,0 +1,262 @@
+	case *persistence.TimeoutError:
+		scopeWithDomainTag.IncCounter(metrics.PersistenceErrTimeoutCounterPerDomain)
+		scopeWithDomainTag.IncCounter(metrics.PersistenceFailuresPerDomain)
+	case *types.ServiceBusyError:
+		scopeWithDomainTag.IncCounter(metrics.PersistenceErrBusyCounterPerDomain)
+		scopeWithDomainTag.IncCounter(metrics.PersistenceFailuresPerDomain)
+	case *persistence.DBUnavailableError:
+		scopeWithDomainTag.IncCounter(metrics.PersistenceErrDBUnavailableCounterPerDomain)
+		scopeWithDomainTag.IncCounter(metrics.PersistenceFailuresPerDomain)
+		p.logger.Error("DBUnavailable Error:", tag.Error(err), tag.MetricScope(scope))
+	default:
+		p.logger.Error("Operation failed with internal error.", tag.Error(err), tag.MetricScope(scope))
+		scopeWithDomainTag.IncCounter(metrics.PersistenceFailuresPerDomain)
+	}
+}
+
+func (p *base) updateErrorMetric(scope int, err error, metricsScope metrics.Scope) {
+	switch err.(type) {
+	case *types.DomainAlreadyExistsError:
+		metricsScope.IncCounter(metrics.PersistenceErrDomainAlreadyExistsCounter)
+	case *types.BadRequestError:
+		metricsScope.IncCounter(metrics.PersistenceErrBadRequestCounter)
+	case *persistence.WorkflowExecutionAlreadyStartedError:
+		metricsScope.IncCounter(metrics.PersistenceErrExecutionAlreadyStartedCounter)
+	case *persistence.ConditionFailedError:
+		metricsScope.IncCounter(metrics.PersistenceErrConditionFailedCounter)
+	case *persistence.CurrentWorkflowConditionFailedError:
+		metricsScope.IncCounter(metrics.PersistenceErrCurrentWorkflowConditionFailedCounter)
+	case *persistence.ShardAlreadyExistError:
+		metricsScope.IncCounter(metrics.PersistenceErrShardExistsCounter)
+	case *persistence.ShardOwnershipLostError:
+		metricsScope.IncCounter(metrics.PersistenceErrShardOwnershipLostCounter)
+	case *types.EntityNotExistsError:
+		metricsScope.IncCounter(metrics.PersistenceErrEntityNotExistsCounter)
+	case *persistence.TimeoutError:
+		metricsScope.IncCounter(metrics.PersistenceErrTimeoutCounter)
+		metricsScope.IncCounter(metrics.PersistenceFailures)
+	case *types.ServiceBusyError:
+		metricsScope.IncCounter(metrics.PersistenceErrBusyCounter)
+		metricsScope.IncCounter(metrics.PersistenceFailures)
+	case *persistence.DBUnavailableError:
+		metricsScope.IncCounter(metrics.PersistenceErrDBUnavailableCounter)
+		metricsScope.IncCounter(metrics.PersistenceFailures)
+		p.logger.Error("DBUnavailable Error:", tag.Error(err), tag.MetricScope(scope))
+	default:
+		p.logger.Error("Operation failed with internal error.", tag.Error(err), tag.MetricScope(scope))
+		metricsScope.IncCounter(metrics.PersistenceFailures)
+	}
+}
+
+func (p *base) call(scope int, op func() error, tags ...metrics.Tag) error {
+	metricsScope := p.metricClient.Scope(scope, tags...)
+	if len(tags) > 0 {
+		metricsScope.IncCounter(metrics.PersistenceRequestsPerDomain)
+	} else {
+		metricsScope.IncCounter(metrics.PersistenceRequests)
```

Comment:
Not related to this change but I don't like this pattern of defining multiple counters for same thing. For example I want to introduce shard-id tag and it doesn't make sense to create a new counter `PersistenceRequestsPerDomainPerShard` :) Ideally we could have a single counter for persistence latency with all the dimensions in it and when a dimension is not applicable it can just be empty.